core/types/calls.cc

```diff
@@ -4061,6 +4061,63 @@ class Array_zip : public IntrinsicMethod {
     }
 } Array_zip;
 
+class Symbol_eqeq : public IntrinsicMethod {
+public:
+    void apply(const GlobalState &gs, const DispatchArgs &args, DispatchResult &res) const override {
+        if (args.args.size() != 1) {
+            return;
+        }
+
+        // NOTE:
+        // If you update this, please update error-reference to mention which types this check applies to
+        auto isOnlySymbol =
+            Types::isSubType(gs, args.fullType.type, Types::any(gs, Types::nilClass(), Types::Symbol()));
+        if (isOnlySymbol && Types::all(gs, args.fullType.type, args.args[0]->type).isBottom()) {
+            auto funLoc = args.funLoc();
+            auto errLoc = (funLoc.exists() && !funLoc.empty()) ? funLoc : args.callLoc();
+            if (auto e = gs.beginError(errLoc, errors::Infer::NonOverlappingEqual)) {
+                e.setHeader("Comparison between `{}` and `{}` is always false", args.fullType.type.show(gs),
+                            args.args[0]->type.show(gs));
+                e.addErrorSection(args.fullType.explainGot(gs, args.originForUninitialized));
+                e.addErrorSection(args.args[0]->explainGot(gs, args.originForUninitialized));
+                if (args.args[0]->type == Types::String() && args.argLoc(0).exists()) {
+                    e.replaceWith("Convert arg to Symbol", args.argLoc(0).copyEndWithZeroLength(), ".to_sym");
+                }
+            }
+        }
+    }
+} Symbol_eqeq;
+
+class String_eqeq : public IntrinsicMethod {
+public:
+    void apply(const GlobalState &gs, const DispatchArgs &args, DispatchResult &res) const override {
+        if (args.args.size() != 1) {
+            return;
+        }
+
+        auto isOnlyString =
+            Types::isSubType(gs, args.fullType.type, Types::any(gs, Types::nilClass(), Types::String()));
+
+        // NOTE:
+        // If you update this, please update error-reference to mention which types this check applies to
+        if (isOnlyString &&
+            Types::isSubType(gs, args.args[0]->type, Types::any(gs, Types::nilClass(), Types::Symbol())) &&
+            Types::all(gs, args.fullType.type, args.args[0]->type).isBottom()) {
+            auto funLoc = args.funLoc();
+            auto errLoc = (funLoc.exists() && !funLoc.empty()) ? funLoc : args.callLoc();
+            if (auto e = gs.beginError(errLoc, errors::Infer::NonOverlappingEqual)) {
+                e.setHeader("Comparison between `{}` and `{}` is always false", args.fullType.type.show(gs),
+                            args.args[0]->type.show(gs));
+                e.addErrorSection(args.fullType.explainGot(gs, args.originForUninitialized));
+                e.addErrorSection(args.args[0]->explainGot(gs, args.originForUninitialized));
+                if (args.args[0]->type == Types::Symbol() && args.receiverLoc().exists()) {
+                    e.replaceWith("Convert arg to Symbol", args.receiverLoc().copyEndWithZeroLength(), ".to_sym");
+                }
```

Comment:
This looks like an overzealous copy-and-paste?  (The `Symbol` case, above, suggests that this wants to be "Convert arg to String" with ".to_s"...but this one is using `receiverLoc()` instead of `argLoc(0)`, so I'm not sure which bits need to be changed.)

Reply:
I think it's correct as is. In a comparison where one is `String` and one is `Symbol`, I wanted to always suggest converting the `String` to a `Symbol`, regardless of whether it was on the left or right of the `==`